clients/src/main/java/org/apache/kafka/clients/consumer/LogTruncationException.java

```diff
@@ -36,7 +36,8 @@
     private final Map<TopicPartition, OffsetAndMetadata> divergentOffsets;
 
     public LogTruncationException(Map<TopicPartition, OffsetAndMetadata> divergentOffsets) {
-        super(Utils.transformMap(divergentOffsets, Function.identity(), OffsetAndMetadata::offset));
+        super(Utils.transformMap(divergentOffsets, Function.identity(), OffsetAndMetadata::offset),
+            "detected log truncation");
```

Comment:
I'd suggest a minor change to this. One of the issues here is that we don't get the divergent offsets in the message itself. I know that might not seem like a big deal, but sometimes the exception trace is all we get from the user. It's also surprising when an exception constructor takes a string parameter which is not the message itself. Can we do something like this instead?
```java
class LogTruncationException {
  public LogTruncationException(Map<TopicPartition, OffsetAndMetadata> divergentOffsets) {
    super(Utils.transformMap(divergentOffsets, Function.identity(), OffsetAndMetadata::offset),
      "Detected log truncation with diverging offsets " + divergentOffsets);
    this.divergentOffsets = Collections.unmodifiableMap(divergentOffsets);
  }
}

class OffsetOutOfRangeException {
  public OffsetOutOfRangeException(Map<TopicPartition, Long> offsetOutOfRangePartitions) {
    this(offsetOutOfRangePartitions, "Offsets out of range with no configured reset policy for partitions: " +
            offsetOutOfRangePartitions);
  }

  public OffsetOutOfRangeException(String message, Map<TopicPartition, Long> offsetOutOfRangePartitions) {
    super(message);
    this.offsetOutOfRangePartitions = offsetOutOfRangePartitions;
  }
}

class Fetcher {
  private handleOffsetOutOfRange(..., String reason) {
    ...
    return new OffsetOutOfRangeException(offsetOutOfRangePartitions, "Offsets out of range with no configured reset policy for partitions: " +
            offsetOutOfRangePartitions + ", root cause: " + reason);
  }
}
```